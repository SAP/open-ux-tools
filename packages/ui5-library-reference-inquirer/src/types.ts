import type { PromptSeverityMessage, YUIQuestion } from '@sap-ux/inquirer-common';
import type { ReuseLib } from '@sap-ux/project-access';
import type { ListChoiceOptions, Validator } from 'inquirer';
/**
 * Enumeration of prompt names used by UI5LibraryReferencePromptOptions
 *
 */
export enum promptNames {
    /**
     * Target project folder
     */
    targetProjectFolder = 'targetProjectFolder',
    /**
     * Source of reuse libraries
     */
    source = 'source',
    /**
     * Libraries to reference
     */
    referenceLibraries = 'referenceLibraries'
}

// Creates a general type for all string value prompt options
type stringValuePromptType = typeof promptNames;
type stringValuePrompts = stringValuePromptType[keyof stringValuePromptType];

/**
 * Defines prompt/question default values and/or whether or not they should be shown.
 */
export type CommonPromptOptions = {
    hide?: boolean;
    validate?: Validator<UI5LibraryReferenceAnswers>;
    additionalMessages?: PromptSeverityMessage;
};

/**
 * Provide the correct type checking for string value prompts and `ui5Version` options
 *
 */
type stringValuePromptOptions = Record<stringValuePrompts, CommonPromptOptions>;

/**
 * Answers for the UI5 Library Reference
 */
export interface UI5LibraryReferenceAnswers {
    targetProjectFolder?: string;
    source?: string;
    referenceLibraries?: ReuseLib[];
}

export type UI5LibraryReferenceQuestion = YUIQuestion<UI5LibraryReferenceAnswers>;

export type UI5LibraryReferencePromptOptions = Partial<stringValuePromptOptions>;

/**
 * Choice for the reuse library
 */
export interface ReuseLibChoice extends ListChoiceOptions {
    name: string;
    value: ReuseLib;
}
