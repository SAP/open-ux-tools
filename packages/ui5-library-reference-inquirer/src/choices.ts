import type { AllAppResults, ReuseLib } from '@sap-ux/project-access';
import type { ListChoiceOptions } from 'inquirer';
import type { ReuseLibChoice } from './types';
import { basename } from 'path';

/**
 * Creates the project choice options.
 *
 * @param apps array of applications found in the workspace folders.
 * @returns list of projects
 */
export const getProjectChoices = async (apps?: AllAppResults[]): Promise<ListChoiceOptions[]> => {
    const projectChoices: ListChoiceOptions[] = [];

    if (apps) {
        for (const app of apps) {
            const name = basename(app.appRoot);
            projectChoices.push({ name, value: { folderName: name, path: app.appRoot } });
        }
    }

    return projectChoices;
};

/**
 * Creates the library choice options.
 *
 * @param libs - array of libraries found in the workspace folders.
 * @returns  list of libraries
 */
export const getLibraryChoices = async (libs?: ReuseLib[]): Promise<ReuseLibChoice[]> => {
    const libraryChoices: ReuseLibChoice[] = [];

    if (libs) {
        for (const lib of libs) {
            const libName = [lib.name, lib.type, lib.description].filter(Boolean).join(' - ');
            libraryChoices.push({ name: libName, value: lib });
        }
    }

    return libraryChoices;
};
