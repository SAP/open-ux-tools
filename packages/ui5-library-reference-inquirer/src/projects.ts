import type { AllAppResults } from '@sap-ux/project-access';
import type { ListChoiceOptions } from 'inquirer';
import { basename } from 'path';

/**
 * Creates the project choice options.
 *
 * @param apps array of applications found in the workspace folders.
 * @returns list of projects
 */
export const getProjectChoices = async (apps?: AllAppResults[]): Promise<ListChoiceOptions[]> => {
    const projectChoices: ListChoiceOptions[] = [];

    if (apps) {
        for (const app of apps) {
            const name = basename(app.appRoot);
            projectChoices.push({ name, value: { folderName: name, path: app.appRoot } });
        }
    }

    return projectChoices;
};
