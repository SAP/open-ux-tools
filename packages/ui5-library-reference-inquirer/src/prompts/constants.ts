export enum SourceType {
    Workspace = 'Workspace'
}
