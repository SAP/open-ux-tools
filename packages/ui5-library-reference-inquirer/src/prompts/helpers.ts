import type { Question } from 'inquirer';
import type {
    CommonPromptOptions,
    promptNames,
    UI5LibraryReferenceAnswers,
    UI5LibraryReferencePromptOptions,
    UI5LibraryReferenceQuestion
} from '../types';
import { extendAdditionalMessages, type validate, type YUIQuestion } from '@sap-ux/inquirer-common';

/**
 * Will remove prompts from the specified prompts based on prompt options
 * Removing prompts is preferable to using `when()` to prevent continuous re-evaluation.
 *
 * @param prompts Keyed prompts object containing all possible prompts
 * @param promptOptions prompt options
 * @returns the updated questions
 */
export function hidePrompts(
    prompts: Record<promptNames, UI5LibraryReferenceQuestion>,
    promptOptions?: UI5LibraryReferencePromptOptions
): UI5LibraryReferenceQuestion[] {
    const questions: UI5LibraryReferenceQuestion[] = [];
    if (promptOptions) {
        Object.keys(prompts).forEach((key) => {
            const promptKey = key as keyof typeof promptNames;
            if (!promptOptions?.[promptKey]?.hide) {
                questions.push(prompts[promptKey]);
            }
        });
    } else {
        questions.push(...Object.values(prompts));
    }
    return questions;
}

/**
 * Extends a validate function.
 *
 * @param question - the question to which the validate function will be applied
 * @param validateFunc - the validate function which will be applied to the question
 * @returns the extended validate function
 */
function extendValidate(
    question: Question,
    validateFunc: validate<UI5LibraryReferenceAnswers>
): validate<UI5LibraryReferenceAnswers> {
    const validate = question.validate;
    return (
        value: unknown,
        previousAnswers?: UI5LibraryReferenceAnswers | undefined
    ): ReturnType<validate<UI5LibraryReferenceAnswers>> => {
        const extVal = validateFunc(value, previousAnswers);
        if (extVal !== true) {
            return extVal;
        }
        return typeof validate === 'function' ? validate(value, previousAnswers) : true;
    };
}

/**
 * Extend the existing prompt property function with the one specified in prompt options or add as new.
 *
 * @param question - the question to which the extending function will be applied
 * @param promptOption - prompt options, containing extending functions
 * @param funcName - the question property (function) name to extend
 * @returns the extended question
 */
function applyExtensionFunction(
    question: YUIQuestion,
    promptOption: CommonPromptOptions,
    funcName: 'validate' | 'additionalMessages'
): YUIQuestion {
    let extendedFunc;

    if (funcName === 'validate' && promptOption.validate) {
        extendedFunc = extendValidate(question, promptOption.validate);
    }

    if (funcName === 'additionalMessages' && promptOption.additionalMessages) {
        extendedFunc = extendAdditionalMessages(question, promptOption.additionalMessages);
    }

    question = Object.assign(question, { [funcName]: extendedFunc });
    return question;
}

/**
 * Updates questions with extensions for specific properties. Only `validate`, `default` and `additionalMessages` are currently supported.
 *
 * @param questions - array of prompts to be extended
 * @param promptOptions - the prompt options possibly containing function extensions
 * @returns - the extended questions
 */
export function extendWithOptions(
    questions: UI5LibraryReferenceQuestion[],
    promptOptions: UI5LibraryReferencePromptOptions
): UI5LibraryReferenceQuestion[] {
    questions.forEach((question) => {
        const promptOptKey = question.name as keyof typeof promptNames;
        const promptOpt = promptOptions[promptOptKey];
        if (promptOpt) {
            const propsToExtend = Object.keys(promptOpt);

            for (const extProp of propsToExtend) {
                if (extProp === 'validate' || extProp === 'additionalMessages') {
                    question = applyExtensionFunction(question, promptOpt, extProp);
                }
            }
        }
    });
    return questions;
}
