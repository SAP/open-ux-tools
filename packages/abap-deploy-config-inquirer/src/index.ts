import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
