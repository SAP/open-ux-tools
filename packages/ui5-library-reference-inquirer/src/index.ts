import { type InquirerAdapter } from '@sap-ux/inquirer-common';
import inquirer, { type Question } from 'inquirer';
import { getQuestions } from './prompts';
import type { UI5LibraryReferenceAnswers, UI5LibraryReferencePromptOptions } from './types';
import {
    getReuseLibs,
    findFioriArtifacts,
    type FioriArtifactTypes,
    type WorkspaceFolder
} from '@sap-ux/project-access';
import { getProjectChoices, getLibraryChoices } from './choices';

/**
 * Get the inquirer prompts for ui5 library inquirer.
 *
 * @param wsFolders - workspace folders to search for fiori artifacts
 * @param promptOptions - options that can control some of the prompt behaviour
 * @returns the prompts used to provide input for ui5 library generation
 */
async function getPrompts(
    wsFolders: readonly WorkspaceFolder[] | string[],
    promptOptions?: UI5LibraryReferencePromptOptions
): Promise<Question<UI5LibraryReferenceAnswers>[]> {
    const options = {
        wsFolders: wsFolders,
        artifacts: ['applications', 'libraries'] as FioriArtifactTypes[]
    };
    const fioriArtifacts = await findFioriArtifacts(options);
    const reuseLibs = await getReuseLibs(fioriArtifacts?.libraries);

    const projectChoices = await getProjectChoices(fioriArtifacts?.applications);
    const libraryChoices = await getLibraryChoices(reuseLibs);

    return getQuestions(projectChoices, libraryChoices, promptOptions);
}

/**
 * Prompt for ui5 library generation inputs.
 *
 * @param wsFolders - workspace folders to search for fiori artifacts
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behaviour
 * @returns the prompt answers
 */
async function prompt(
    wsFolders?: readonly WorkspaceFolder[] | string[],
    adapter?: InquirerAdapter,
    promptOptions?: UI5LibraryReferencePromptOptions
): Promise<UI5LibraryReferenceAnswers> {
    const ui5LibPrompts = await exports.getPrompts(promptOptions, wsFolders);

    return adapter ? adapter.prompt(ui5LibPrompts) : inquirer.prompt(ui5LibPrompts);
}

export {
    getPrompts,
    prompt,
    type UI5LibraryReferencePromptOptions,
    type UI5LibraryReferenceAnswers,
    type InquirerAdapter
};
