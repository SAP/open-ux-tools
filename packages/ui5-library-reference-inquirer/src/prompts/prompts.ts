import { t } from '../i18n';
import type { ListChoiceOptions } from 'inquirer';
import { checkDependencies } from '@sap-ux/project-access';
import type { CheckBoxQuestion, ListQuestion } from '@sap-ux/inquirer-common';
import {
    promptNames,
    type UI5LibraryReferenceQuestion,
    type ReuseLibChoice,
    type UI5LibraryReferenceAnswers,
    type UI5LibraryReferencePromptOptions
} from '../types';
import { extendWithOptions, hidePrompts } from './helpers';
import { SourceType } from './constants';
import { Severity, type IMessageSeverity } from '@sap-devx/yeoman-ui-types';

/**
 * Get the prompts for the UI5 library reference writing.
 *
 * @param projectChoices - workspace projects
 * @param reuseLibs - reuse libraries
 * @param promptOptions - optional inputs used to pre-populate some prompt choices, default values and other prompting options
 * @returns the prompts
 */
export function getQuestions(
    projectChoices?: ListChoiceOptions[],
    reuseLibs?: ReuseLibChoice[],
    promptOptions?: UI5LibraryReferencePromptOptions
): UI5LibraryReferenceQuestion[] {
    const keyedPrompts: Record<promptNames, UI5LibraryReferenceQuestion> = {
        [promptNames.targetProjectFolder]: getTargetProjectFolderPrompt(projectChoices),
        [promptNames.source]: getSourcePrompt(),
        [promptNames.referenceLibraries]: getReferenceLibrariesPrompt(reuseLibs)
    };

    // Hide not applicable prompts based on passed options
    let questions: UI5LibraryReferenceQuestion[] = hidePrompts(keyedPrompts, promptOptions);

    // Apply extended `validate`, `additionalMessages` or override `default` prompt properties
    if (promptOptions) {
        questions = extendWithOptions(questions, promptOptions);
    }

    return questions;
}

/**
 * Get the target project folder prompt.
 *
 * @param projectChoices workspace projects
 * @returns The `targetProjectFolder` prompt
 */
function getTargetProjectFolderPrompt(projectChoices?: ListChoiceOptions[]): UI5LibraryReferenceQuestion {
    return {
        type: 'list',
        guiOptions: {
            breadcrumb: true
        },
        name: promptNames.targetProjectFolder,
        message: t('prompts.targetProjectFolderLabel'),
        choices: projectChoices,
        default: () => {
            return projectChoices?.length ? 0 : undefined;
        },
        validate: () => {
            return !projectChoices?.length ? t('error.noProjectsFound') : true;
        }
    } as ListQuestion<UI5LibraryReferenceAnswers>;
}

/**
 * Get the source prompt.
 *
 * @returns The `source` prompt
 */
function getSourcePrompt(): UI5LibraryReferenceQuestion {
    return {
        type: 'list',
        guiOptions: {
            breadcrumb: true
        },
        name: promptNames.source,
        message: t('prompts.sourceLabel'),
        choices: [{ name: t('choices.workspace'), value: SourceType.Workspace }],
        default: 0
    } as ListQuestion<UI5LibraryReferenceAnswers>;
}

/**
 * Get the reference libraries prompt.
 *
 * @param reuseLibs reuse libraries
 * @returns The `referenceLibraries` prompt
 */
function getReferenceLibrariesPrompt(reuseLibs?: ReuseLibChoice[]): UI5LibraryReferenceQuestion {
    let missingDeps: string;
    return {
        type: 'checkbox',
        name: promptNames.referenceLibraries,
        message: t('prompts.referenceLibrariesLabel'),
        guiOptions: {
            breadcrumb: true
        },
        choices: reuseLibs,
        additionalMessages: (): IMessageSeverity | undefined => {
            const msg = {
                message: t('addtionalMsgs.missingDeps', { dependencies: missingDeps }),
                severity: Severity.warning
            };
            return missingDeps ? msg : undefined;
        },
        validate: (answer) => {
            if (!reuseLibs?.length) {
                return t('error.noLibsFound');
            } else if (answer?.length < 1) {
                return t('error.noLibSelected');
            } else if (answer?.length) {
                missingDeps = checkDependencies(
                    answer,
                    reuseLibs.map((libs) => libs.value)
                );
            }
            return true;
        }
    } as CheckBoxQuestion<UI5LibraryReferenceAnswers>;
}
