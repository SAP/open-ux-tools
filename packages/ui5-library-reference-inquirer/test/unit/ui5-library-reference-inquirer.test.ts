import { join } from 'path';
import { prompt, getPrompts, type InquirerAdapter, type UI5LibraryReferencePromptOptions } from '../../src';
import * as ui5LibRefPrompts from '../../src/prompts';
import { createPromptModule } from 'inquirer';
import { promptNames } from '../../src/types';
import * as uxProjectAccess from '@sap-ux/project-access';

jest.mock('../../src/prompts', () => {
    return {
        __esModule: true,
        ...jest.requireActual('../../src/prompts')
    };
});

describe('ui5-library-reference-inquirer API', () => {
    jest.setTimeout(1000000000);
    afterEach(() => {
        jest.restoreAllMocks();
    });

    test('getPrompts, no options', async () => {
        const getQuestionsSpy = jest.spyOn(ui5LibRefPrompts, 'getQuestions');
        // All prompts, no options
        const prompts = await getPrompts([join(__dirname, '/samples')]);
        expect(prompts).toMatchSnapshot();
        expect(getQuestionsSpy).toHaveBeenCalled();
    });

    test('prompt, prompt module registers plugin', async () => {
        const mockPromptsModule = createPromptModule();
        jest.spyOn(uxProjectAccess, 'findFioriArtifacts').mockResolvedValue([] as uxProjectAccess.FoundFioriArtifacts);
        const mockInquirerAdapter: InquirerAdapter = {
            prompt: jest.fn().mockResolvedValue({ aPrompt: 'a prompt answer' }),
            promptModule: mockPromptsModule
        };
        const promptOpts: UI5LibraryReferencePromptOptions = {
            [promptNames.source]: {
                hide: true
            }
        };

        expect(await prompt([join(__dirname, '/samples'), 'test'], mockInquirerAdapter, promptOpts))
            .toMatchInlineSnapshot(`
            Object {
              "aPrompt": "a prompt answer",
            }
        `);
    });
});
