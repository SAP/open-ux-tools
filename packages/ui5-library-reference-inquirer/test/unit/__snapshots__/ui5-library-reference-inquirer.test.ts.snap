// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui5-library-reference-inquirer API getPrompts, no options 1`] = `
Array [
  Object {
    "choices": Array [],
    "default": [Function],
    "guiOptions": Object {
      "breadcrumb": true,
    },
    "message": "Project folder path",
    "name": "targetProjectFolder",
    "type": "list",
    "validate": [Function],
  },
  Object {
    "choices": Array [
      Object {
        "name": "Workspace",
        "value": "Workspace",
      },
    ],
    "default": 0,
    "guiOptions": Object {
      "breadcrumb": true,
    },
    "message": "Reusable Library Source",
    "name": "source",
    "type": "list",
  },
  Object {
    "additionalMessages": [Function],
    "choices": Array [],
    "guiOptions": Object {
      "breadcrumb": true,
    },
    "message": "Select Reusable Libraries",
    "name": "referenceLibraries",
    "type": "checkbox",
    "validate": [Function],
  },
]
`;
